src/source-af-xdp.c

```diff
@@ -0,0 +1,944 @@
+
+    /* Has the eBPF program successfully bound? */
+    if (bpf_get_link_xdp_id(ptv->ifindex, &ptv->prog_id, ptv->xsk.cfg.xdp_flags)) {
+        SCLogError(SC_ERR_BPF, "Failed to attach eBPF program to interface: %s", ptv->livedev->dev);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+/**
+ * \brief Try to reopen AF_XDP socket
+ *
+ * \retval: TM_ECODE_OK in case of success
+ * TM_ECODE_FAILED if error occurs or a condition is not met.
+ */
+static TmEcode AFXDPTryReopen(AFXDPThreadVars *ptv)
+{
+    AFXDPCloseSocket(ptv);
+    usleep(RECONNECT_TIMEOUT);
+
+    /* open (any) socket and check if state = UP or DOWN */
+    int sock_fd = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
+    if (sock_fd == -1) {
+        SCLogError(SC_ERR_AFXDP_CREATE, "Couldn't create a AF_PACKET socket, error %s",
+                strerror(errno));
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    int if_flags = AFXDPGetDevFlags(sock_fd, ptv->iface);
+    if (if_flags == -1) {
+        SCLogDebug("Couldn't get flags for interface '%s'", ptv->iface);
+        goto sock_err;
+    } else if ((if_flags & (IFF_UP | IFF_RUNNING)) == 0) {
+        SCLogDebug("Interface '%s' is down", ptv->iface);
+        goto sock_err;
+    }
+    close(sock_fd);
+
+    if (AFXDPSocketCreation(ptv) != TM_ECODE_OK) {
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    SCLogInfo("Interface '%s' is back", ptv->iface);
+    SCReturnInt(TM_ECODE_OK);
+
+sock_err:
+    close(sock_fd);
+    SCReturnInt(TM_ECODE_FAILED);
+}
+
+static void AFXDPReleasePacket(Packet *p)
+{
+    *xsk_ring_prod__fill_addr((struct xsk_ring_prod *)p->afxdp_v.fq, p->afxdp_v.fq_idx) =
+            p->afxdp_v.orig;
+
+    PacketFreeOrRelease(p);
+}
+
+static inline int DumpStatsEverySecond(AFXDPThreadVars *ptv, time_t *last_dump)
+{
+    int stats_dumped = 0;
+    time_t current_time = time(NULL);
+
+    if (current_time != *last_dump) {
+        AFXDPDumpCounters(ptv);
+        *last_dump = current_time;
+        stats_dumped = 1;
+    }
+
+    StatsSyncCountersIfSignalled(ptv->tv);
+
+    return stats_dumped;
+}
+
+static inline ssize_t WakeupSocket(void *data)
+{
+    ssize_t res = 0;
+    AFXDPThreadVars *ptv = (AFXDPThreadVars *)data;
+
+    /* Assuming kernel >= 5.11 in use if xdp_busy_poll is enabled */
+    if (ptv->xsk.enable_busy_poll || xsk_ring_prod__needs_wakeup(&ptv->umem.fq)) {
+        res = recvfrom(xsk_socket__fd(ptv->xsk.xsk), NULL, 0, MSG_DONTWAIT, NULL, NULL);
+    }
+
+    return res;
+}
+
+/**
+ *  \brief Main AF_XDP reading Loop function
+ */
+static TmEcode ReceiveAFXDPLoop(ThreadVars *tv, void *data, void *slot)
+{
+    SCEnter();
+
+    Packet *p;
+    time_t last_dump = 0;
+    struct timeval ts;
+    uint32_t idx_rx = 0, idx_fq = 0, rcvd;
+    int r;
+    AFXDPThreadVars *ptv = (AFXDPThreadVars *)data;
+    TmSlot *s = (TmSlot *)slot;
+
+    ptv->slot = s->slot_next;
+
+    /* Reserve memory for umem  */
+    if (AcquireBuffer(ptv) != TM_ECODE_OK) {
```

Comment:
thread specific setup before the loop belongs in `ReceiveAFXDPThreadInit`, any reason to have it here?